### ISSUE TITLE
Segmentation fault when performing operations on operators

### ISSUE THREAD

**Issue by PhilipVinc** [netket/netket#206]
As pointed out by @femtobit in #203, when we perform arithmetic operations with operators in python, creating new objects, we do not tell the reference counter in python that there are new references to the `hilbert` object. This can lead to segmentation faults (See example below):

```python
import netket as nk

g = nk.graph.Hypercube(8, 1)
hi = nk.hilbert.Spin(g, 1/2)

lo = nk.operator.LocalOperator(hi, [[1,0],[0,1]], [0])
lo = lo*lo

# Assign hi to something else so it will no longer keep alive the original Hilbert object
hi = nk.hilbert.Qubit(g)

# This will try to access the now deleted Hilbert object
lo = lo * lo
```

One possible solution might be to use `std::shared_ptr` in place of a reference. Otherwise we must find a way to explicitly update python/pybind reference counter.

**Comment by gcarleo** [comment]
Alright, there are basically two options: 

1. Use shared_ptr to handle hilbert inside `Operator`s 
2. Make `Hilbert` a concrete object and copy it around 

I would honestly prefer option 2 but one has to think a little bit about the requirements to do that without compromising the interface we have now

**Comment by gcarleo** [comment]
I think we need to do a clone method in the abstract class, and store
Hilbert as a unique_ptr. This has minimal impact ont he code.

**Comment by PhilipVinc** [comment]
So as you said there are two solutions and it revolves around what do we want to do with `Hilbert`. Key point is "who is the owner of each object".

Currently, in `netket` we store references to `Hilbert` and `Graph`, which means that objects (such as `Operators`) are not the owners of `Hilbert` and `Graph`, rather, the python runtime is. And this causes this segmentation fault because python's Garbage Collector is not aware of this.

To be more explicit about object ownership we can:
 2 - Store `Hilbert` by value in `Operator`, and triggering a `copy` of `Hilbert` every time we create a new operator. There are two problems with this approach: 
  - (i) you can only store by value concrete objects, and this would remove flexibility from the design point of view (or we need to make everything a template, which I don't think it's a good idea) 
  - (ii) when we call `GetHilbert` from python we must return a copy of `Hilbert` and not a reference, otherwise we hit this bug again
  - (iii) to improve performance inside of the c++ core, we should add a method `GetHilbertNoCopy` returning a `const AbstractHilbert&&` so that we return a const r-value that can't be binded to any variable. This would essentially behave as before (returning a reference) but it checks explicitly at the compiler level that you are not binding the return value.  


1 - Store `Hilbert` as a `unique_ptr<AbstractHilbert>` both in c++ and python. You can store the `unique_ptr` to `Hilbert` in AbstractOperator as a public object, so that you access it directly and not with `GetHilbert`. All constructors should copy-construct starting from this. Basically it's the same as 2 from the point of view of Ownership, but with pointers instead of values.

1b - Store `Hilbert` as a (public) `shared_ptr<AbstractHilbert>`. We should change the `pybind` bindings to store `Hilbert` as a shared_ptr. All methods that accept `Hilbert` should be updated to take a `shared_ptr`.

--

Solution 2) is either overly complicated (templates) or not generic (values) so I would exclude it. With respect to smart_pointers, it does not change much, but probably using `shared_ptr` would require the least code changes. *Note* that if we go down this route, we should remove all usages of references from the code, otherwise this defeats the purpose of using `shared_ptr`

**Comment by femtobit** [comment]
Should be fixed by #207
